include/clang/AST/Expr.h

```diff
@@ -2798,7 +2798,9 @@ class BoundsExpr : public Expr {
 /// classes).
 class CastExpr : public Expr {
 private:
-  enum { OP, BOUNDS, END_EXPR = 2 };
+  // BOUNDS - enum value for bounds of cast expression
+  // SUBBOUNDS - enum value for bounds of subexpression
+  enum { OP, BOUNDS, SUBBOUNDS, END_EXPR = 3 };
```

Comment:
Instead of overwriting the pointer to the existing bounds expression, I would like to have a new pointer that is set during the semantic analysis phase.    I suggest adding another entry to the SubExprs array for this pointer.